linux/ibus-keyman/src/KeymanSystemServiceClient.cpp

```diff
@@ -0,0 +1,97 @@
+#include <systemd/sd-bus.h>
+#include "KeymanSystemServiceClient.h"
+
+#define KEYMAN_BUS_NAME "com.keyman.SystemService1"
+#define KEYMAN_INTERFACE_NAME "com.keyman.SystemService1.System"
+#define KEYMAN_OBJECT_PATH "/com/keyman/SystemService1/System"
+
+extern gboolean testing;
+
+class KeymanSystemServiceClient {
+private:
+  sd_bus_error *error = NULL;
+  sd_bus_message *msg = NULL;
+  sd_bus *bus         = NULL;
+
+public:
+  KeymanSystemServiceClient();
+  ~KeymanSystemServiceClient();
+
+  void SetCapsLockIndicator(guint32 capsLock);
+  gint32 GetCapsLockIndicator();
+};
+
+KeymanSystemServiceClient::KeymanSystemServiceClient() {
+  int result;
+  if (testing) {
+    result = sd_bus_open_user(&bus);
+  } else {
+    result = sd_bus_default_system(&bus);
+  }
+  if (result < 0) {
+    g_error("%s: Can't get connection: %s", __FUNCTION__, strerror(-result));
+  }
+}
+
+KeymanSystemServiceClient::~KeymanSystemServiceClient() {
+  if (error) { sd_bus_error_free(error); }
+  if (msg)   { sd_bus_message_unref(msg); }
+  if (bus)   { sd_bus_unref(bus); }
+}
```

Comment:
In the pattern you have now, if you call `SetCapsLockIndicator` twice and both times an error is returned, you'll leak `error`. It would be good to make note of this -- that the class functions are intended to be called just once in the class lifecycle, for future maintainers.